pkg/controller/certificatemanager/certificatemanager.go

```diff
@@ -237,44 +240,84 @@ func (cm *certificateManager) getKeyPair(cli client.Client, secretName, secretNa
 		return nil, nil, err
 	}
 
+	// First, check that it's not expired.
 	if x509Cert.NotAfter.Before(time.Now()) || x509Cert.NotBefore.After(time.Now()) {
 		if !readCertOnly && strings.HasPrefix(x509Cert.Issuer.CommonName, rmeta.TigeraOperatorCAIssuerPrefix) {
 			if cm.keyPair.CertificateManagement != nil {
 				// When certificate management is enabled, we can simply return a certificate management key pair;
 				// the old secret will be deleted automatically.
 				return certificateManagementKeyPair(cm, secretName, nil), nil, nil
 			}
+
 			// We return nil, so a new secret will be created for expired (legacy) operator signed secrets.
 			return nil, nil, nil
 		}
-		// We return an error for byo secrets.
+
+		// The certificate was not issued by tigera-operator, or the caller indicated that certs shouldn't be modified.
+		// Consider this a BYO certificate, and error out rather than generate a new certificate.
 		return nil, nil, fmt.Errorf("secret %s is not valid at this date", secretName)
 	}
 	var issuer certificatemanagement.KeyPairInterface
 	if x509Cert.Issuer.CommonName == rmeta.TigeraOperatorCAIssuerPrefix {
 		if cm.keyPair.CertificateManagement != nil {
 			return certificateManagementKeyPair(cm, secretName, nil), nil, nil
 		}
+
+		// This certificate wasn't issued using external certificate management.
+		if !ValidForClientAndServer(x509Cert) {
+			// The certificate isn't valid as both a client and server cert. Older versions of the operator
+			// generated certificates that were only valid for use as a server certificate. However, some components now
+			// use these as client certificates in mTLS connections as well. So, regenerate any that need this capability.
+			return nil, nil, nil
+		}
+
 		if string(x509Cert.AuthorityKeyId) == string(cm.AuthorityKeyId) {
+			// The certifiate's authority matches the current CA.
 			issuer = cm.keyPair
 		} else {
+			// The certificate's authority doesn't match the current CA.
 			if !readCertOnly {
-				// We want to return nothing, so a new secret will be created to overwrite this one.
+				// The caller indicated we're allowed to generate new certs - return nothing, so
+				// that a new secret will be created to overwrite this one.
 				return nil, nil, nil
 			}
+
 			// We treat the certificate as a BYO secret, because this may be a certificate created by a management cluster
 			// and used inside a managed cluster. If it is not, it should get updated automatically when readCertOnly=false.
 			issuer = nil
 		}
 	}
+
 	return &certificatemanagement.KeyPair{
 		Issuer:         issuer,
 		Name:           secretName,
 		PrivateKeyPEM:  keyPEM,
 		CertificatePEM: certPEM,
 		OriginalSecret: secret,
 	}, x509Cert, nil
+}
+
+func X509FromSecret(secret *corev1.Secret) (*x509.Certificate, error) {
+	_, certPEM := getKeyCertPEM(secret)
+	x509Cert, err := certificatemanagement.ParseCertificate(certPEM)
+	if err != nil {
+		return nil, err
+	}
+	return x509Cert, nil
+}
```

Comment:
Would it be better to define this in a test-only code file, as IIUC it is only used for UT?

Reply:
I would have liked to, although the specific functionality that I wanted from here was the `getKeyCertPEM` which is currently private.

I suppose I could just make that public though!

Reply:
Ah I see - whatever you prefer then!